src/v/cloud_storage/tests/delete_records_e2e_test.cc

```diff
@@ -311,3 +311,177 @@ FIXTURE_TEST(test_delete_from_archive_consume, delete_records_e2e_fixture) {
+    auto& stm_manifest = archiver->manifest();
+    BOOST_REQUIRE_EQUAL(stm_manifest.get_spillover_map().size(), 3);
+    BOOST_REQUIRE_EQUAL(stm_manifest.size(), segs_per_spill);
+
+    // DeleteRecords to just before the end of the local log.
+    kafka_delete_records_transport deleter(make_kafka_client().get());
+    deleter.start().get();
+    auto end_offset = kafka::offset(245);
+    BOOST_REQUIRE_EQUAL(
+      end_offset,
+      deleter
+        .delete_records_from_partition(
+          topic_name,
+          model::partition_id(0),
+          kafka::offset_cast(end_offset),
+          5s)
+        .get());
+
+    // The first housekeeping should remove all spillover segments.
+    auto archive_start = stm_manifest.get_archive_start_offset();
+    auto stm_start_before = stm_manifest.get_start_offset();
+    BOOST_REQUIRE_EQUAL(archive_start, model::offset(0));
+    archiver->housekeeping().get();
+    BOOST_REQUIRE_EQUAL(
+      stm_manifest.get_archive_start_offset(), model::offset{});
+    BOOST_REQUIRE_EQUAL(10, archiver->manifest().size());
+    BOOST_REQUIRE_EQUAL(stm_start_before, stm_manifest.get_start_offset());
+
+    // The next should remove all STM segments.
+    archiver->housekeeping().get();
+    BOOST_REQUIRE(!archiver->manifest().get_start_offset().has_value());
+}
+
+// Test that truncation is applied to cloud storage as expected when a
+// DeleteRecords request lands in the STM manifest.
+FIXTURE_TEST(test_delete_from_stm_truncation, delete_records_e2e_fixture) {
+    tests::remote_segment_generator gen(make_kafka_client().get(), *partition);
+    size_t records_per_seg = 5;
+    BOOST_REQUIRE_GE(
+      250,
+      gen.batches_per_segment(records_per_seg)
+        .num_segments(40)
+        .additional_local_segments(10)
+        .produce()
+        .get());
+    archiver->apply_spillover().get();
+    auto& stm_manifest = archiver->manifest();
+    BOOST_REQUIRE_EQUAL(stm_manifest.get_spillover_map().size(), 3);
+    BOOST_REQUIRE_EQUAL(stm_manifest.size(), segs_per_spill);
+    kafka_delete_records_transport deleter(make_kafka_client().get());
+    deleter.start().get();
+
+    // Truncate within the STM at the bound, leaving one segment.
+    auto stm_truncate_offset
+      = archiver->manifest().get_last_kafka_offset().value();
+    deleter
+      .delete_records_from_partition(
+        topic_name,
+        model::partition_id(0),
+        kafka::offset_cast(stm_truncate_offset),
+        5s)
+      .get();
+    // The first housekeeping will cleanup the archive.
+    archiver->housekeeping().get();
+    BOOST_REQUIRE_EQUAL(10, archiver->manifest().size());
+    BOOST_REQUIRE_EQUAL(
+      stm_manifest.get_archive_start_offset(), model::offset{});
+
+    // The next will clear the STM manifest.
+    archiver->housekeeping().get();
+    BOOST_REQUIRE_EQUAL(1, archiver->manifest().size());
+    BOOST_REQUIRE_EQUAL(
+      stm_manifest.get_archive_start_offset(), model::offset{});
```

Comment:
suggestion: can we also check that `_start_kafka_offset` in the partition's manifest has been reset? Ditto for all other tests.